crates/optimism/primitives/src/op_tx_type.rs

```diff
@@ -0,0 +1,197 @@
+//! newtype pattern on op_alloy_consensus::OpTxType.
+//! OpTxType that implements reth_primitives_traits::TxType.
+//! This type is required because a Compact impl is needed on the deposit tx type.
+
+use alloy_eips::eip2718::Eip2718Error;
+use alloy_primitives::{U64, U8};
+use alloy_rlp::{Decodable, Encodable};
+use bytes::BufMut;
+use core::fmt::Debug;
+use derive_more::Display;
+use op_alloy_consensus::OpTxType as AlloyOpTxType;
+use reth_primitives_traits::TxType;
+use std::convert::TryFrom;
+
+/// Wrapper type for `AlloyOpTxType` to implement `TxType` trait.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Display, Ord, Hash)]
+pub struct OpTxType(AlloyOpTxType);
+
+impl From<AlloyOpTxType> for OpTxType {
+    fn from(tx_type: AlloyOpTxType) -> Self {
+        OpTxType(tx_type)
+    }
+}
+
+impl From<OpTxType> for u8 {
+    fn from(tx_type: OpTxType) -> Self {
+        tx_type.0 as u8
+    }
+}
+
+impl From<OpTxType> for U8 {
+    fn from(tx_type: OpTxType) -> Self {
+        U8::from(u8::from(tx_type))
+    }
+}
+
+impl TryFrom<u8> for OpTxType {
+    type Error = Eip2718Error;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        AlloyOpTxType::try_from(value)
+            .map(OpTxType)
+            .map_err(|_| Eip2718Error::UnexpectedType(value))
+    }
+}
+
+impl Default for OpTxType {
+    fn default() -> Self {
+        // TODO: default to `Deposit` for optimism ?
+        OpTxType(AlloyOpTxType::Deposit)
+    }
+}
+
+impl PartialEq<u8> for OpTxType {
+    fn eq(&self, other: &u8) -> bool {
+        let self_as_u8: u8 = (*self).into();
+        &self_as_u8 == other
+    }
+}
+
+impl TryFrom<u64> for OpTxType {
+    type Error = Eip2718Error;
+
+    fn try_from(value: u64) -> Result<Self, Self::Error> {
+        if value > u8::MAX as u64 {
+            return Err(Eip2718Error::UnexpectedType(0));
+        }
+        Self::try_from(value as u8)
+    }
+}
+
+impl TryFrom<U64> for OpTxType {
+    type Error = Eip2718Error;
+
+    fn try_from(value: U64) -> Result<Self, Self::Error> {
+        let u64_value: u64 = value.try_into().map_err(|_| Eip2718Error::UnexpectedType(0))?;
+        Self::try_from(u64_value)
+    }
+}
```

Comment:
can also use derive more instead

Reply:
same as above, I don't know a way to derive `TryFrom` via macros for Structs